{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(fname):\n",
    "    headers, records = [], []\n",
    "    record = \"\"\n",
    "    with open(fname, \"rb\") as fin:\n",
    "        for line in fin:\n",
    "            if not line[0] == '>':\n",
    "                record += line.strip()\n",
    "            else:\n",
    "                headers.append(line.strip())\n",
    "                if record:\n",
    "                    records.append(record)\n",
    "                    record = \"\"\n",
    "    if record:\n",
    "        records.append(record)\n",
    "    return headers, records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CAC': 1, 'ACA': 3}\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "def get_frames(genome):\n",
    "    frames, n = [], len(genome)\n",
    "    for start in range(3):\n",
    "        frame = []\n",
    "        for i in range(start, n, 3):\n",
    "            if (i + 3) <= n:\n",
    "                frame.append(genome[i:i + 3])\n",
    "        frames.append(frame)\n",
    "    return frames\n",
    "\n",
    "def get_orfs(genome):\n",
    "    START_CODON = \"ATG\"\n",
    "    END_CODONS = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "    orfs = []\n",
    "    start = -1\n",
    "    while start < len(genome):\n",
    "        start = genome.find(START_CODON, start + 1)\n",
    "        if start < 0:\n",
    "            break\n",
    "        for i in range(start + 3, len(genome), 3):\n",
    "            if genome[i:i+3] in END_CODONS:\n",
    "                orfs.append((start, genome[start:i+3]))\n",
    "                break\n",
    "    return orfs\n",
    "\n",
    "def get_repeats(genome, n):\n",
    "    \"\"\"\n",
    "    n - length of the repeats\n",
    "    \"\"\"\n",
    "    start = -1\n",
    "    repeats = {}\n",
    "    for i in range(len(genome) - n + 1):\n",
    "        pos = i\n",
    "        repeats[genome[i:i+n]].add() = repeats.get(genome[i:i+n], 0) + 1\n",
    "        while pos < len(genome):\n",
    "            pos = genome.find(genome[i:i + n], pos + 1)\n",
    "            if pos >= 0:\n",
    "                repeats[genome[i:i+n]] += 1\n",
    "            else:\n",
    "                break\n",
    "    print repeats\n",
    "print get_repeats(\"ACACA\", 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gi|142022655|gb|EQ086233.1|43 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 990\n",
      "max orf length: 213\n",
      "max orf start: 366\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|160 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 724\n",
      "max orf length: 363\n",
      "max orf start: 106\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|41 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 3080\n",
      "max orf length: 918\n",
      "max orf start: 1193\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|221 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 2863\n",
      "max orf length: 594\n",
      "max orf start: 1771\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|294 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 3832\n",
      "max orf length: 1608\n",
      "max orf start: 140\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|323 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 4805\n",
      "max orf length: 1686\n",
      "max orf start: 2823\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|564 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 1663\n",
      "max orf length: 507\n",
      "max orf start: 824\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|521 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 512\n",
      "max orf length: 159\n",
      "max orf start: 125\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|455 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 691\n",
      "max orf length: 552\n",
      "max orf start: 85\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|229 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 3072\n",
      "max orf length: 1311\n",
      "max orf start: 124\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|422 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 1801\n",
      "max orf length: 639\n",
      "max orf start: 215\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|384 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 3603\n",
      "max orf length: 720\n",
      "max orf start: 1346\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|280 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 2478\n",
      "max orf length: 291\n",
      "max orf start: 1426\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|158 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 1608\n",
      "max orf length: 1218\n",
      "max orf start: 149\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|59 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 4745\n",
      "max orf length: 1107\n",
      "max orf start: 3591\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|319 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 1810\n",
      "max orf length: 537\n",
      "max orf start: 218\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|438 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 3424\n",
      "max orf length: 477\n",
      "max orf start: 986\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|210 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 1451\n",
      "max orf length: 564\n",
      "max orf start: 409\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|237 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 3276\n",
      "max orf length: 219\n",
      "max orf start: 793\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|507 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 2124\n",
      "max orf length: 741\n",
      "max orf start: 929\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|350 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 1712\n",
      "max orf length: 678\n",
      "max orf start: 364\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|245 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 1325\n",
      "max orf length: 636\n",
      "max orf start: 605\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|279 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 1189\n",
      "max orf length: 153\n",
      "max orf start: 412\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|378 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 555\n",
      "max orf length: 534\n",
      "max orf start: 1\n",
      "*****************\n",
      ">gi|142022655|gb|EQ086233.1|101 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence\n",
      "record length: 2449\n",
      "max orf length: 639\n",
      "max orf start: 1632\n",
      "*****************\n"
     ]
    }
   ],
   "source": [
    "headers, records = read_fasta(\"dna.example.fasta\")\n",
    "for i in range(len(headers)):\n",
    "    print headers[i]\n",
    "    print 'record length:', len(records[i])\n",
    "    orfs = get_orfs(records[i])\n",
    "    mlen, mstart = 0, 0\n",
    "    for start, orf in orfs:\n",
    "        if len(orf) > mlen:\n",
    "            mstart = start\n",
    "            mlen = len(orf)\n",
    "    print 'max orf length:', mlen\n",
    "    print 'max orf start:', mstart\n",
    "    print '*****************'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
