As part of the effort to catalog genetic variation in the plant Arabidopsis thaliana, you re-sequenced the genome of one strain ("wu_0_A"; genome file: `wu_0.v7.fas`), to determine genetic variants in this organism. The sequencing reads produced are in the file `wu_0_A_wgs.fastq`. Using the tools bowtie2, samtools and bcftools, develop a pipeline for variant calling in this genome. NOTE: Genome and re-sequencing data have been obtained and modified from those generated by the 1001 Genomes Project, accession ‘Wu_0_A’.

## Apply to questions 1 - 5:
Generate a bowtie2 index of the wu_0_A genome using bowtie2-build, with the prefix 'wu_0'.
```console
$ mkdir athal
$ bowtie2-build wu_0.v7.fas athal/wu_0
```

1. How many sequences were in the genome?
```console
$ grep ">" wu_0.v7.fas | wc -l
7
```

2. What was the name of the third sequence in the genome file? Give the name only, without the “>” sign.
```console
$ grep ">" wu_0.v7.fas | head -n3
>Chr1
>Chr2
>Chr3
# answer: Chr3
```

3. What was the name of the last sequence in the genome file? Give the name only, without the “>” sign.
```console
$ grep ">" wu_0.v7.fas | tail -n1
>mitochondria
```

4. How many index files did the operation create?
```console
6
```

5. What is the 3-character extension for the index files created?
```console
bt2
```


## Apply to questions 6 - 14:
Run bowtie2 to align the reads to the genome, under two scenarios: first, to report only full-length matches of the reads; and second, to allow partial (local) matches. All other parameters are as set by default.
```console
$ bowtie2 -x athal/wu_0 --end-to-end wu_0_A_wgs.fastq -S align.full.sam
147354 reads; of these:
  147354 (100.00%) were unpaired; of these:
    9635 (6.54%) aligned 0 times
    93780 (63.64%) aligned exactly 1 time
    43939 (29.82%) aligned >1 times
93.46% overall alignment rate
$ bowtie2 -x athal/wu_0 --local wu_0_A_wgs.fastq -S align.local.sam
147354 reads; of these:
  147354 (100.00%) were unpaired; of these:
    6310 (4.28%) aligned 0 times
    84939 (57.64%) aligned exactly 1 time
    56105 (38.07%) aligned >1 times
95.72% overall alignment rate
```

6. How many reads were in the original fastq file?
```console
147354
```

7. How many matches (alignments) were reported for the original (full-match) setting? Exclude lines in the file containing unmapped reads.
```console
$ samtools view -F 0x4 -c align.full.sam
137719
```

8. How many matches (alignments) were reported with the local-match setting? Exclude lines in the file containing unmapped reads.
```console
$ samtools view -F 0x4 -c align.local.sam
141044
```

9. How many reads were mapped in the scenario in Question 7?
```console
$ samtools view -F 0x4 align.full.sam | cut -f1 | sort | uniq | wc -l
137719
```

10. How many reads were mapped in the scenario in Question 8?
```console
$ samtools view -F 0x4 align.local.sam | cut -f1 | sort | uniq | wc -l
141044
```

11. How many reads had multiple matches in the scenario in Question 7? You can find this in the bowtie2 summary; note that by default bowtie2 only reports the best match for each read.​
```console
43939
```

12. How many reads had multiple matches in the scenario in Question 8? Use the format above. You can find this in the bowtie2 summary; note that by default bowtie2 only reports the best match for each read.​
```console
56105
```

13. How many alignments contained insertions and/or deletions, in the scenario in Question 7?
```console
$ samtools view -F 0x4 align.full.sam | cut -f6 | grep "[I|D]" | wc -l
2782
```

14. How many alignments contained insertions and/or deletions, in the scenario in Question 8?
```console
$ samtools view -F 0x4 align.local.sam | cut -f6 | grep "[I|D]" | wc -l
2614
```

**For the following set of questions (11 - 20), use the set of full-length alignments calculated under scenario 1 only. Convert this SAM file to BAM, then sort the resulting BAM file.**

```console
$ samtools view align.full.sam -b -o align.full.bam
$ samtools sort -T sorted -o align.full.sorted.bam align.full.bam
$ samtools index align.full.sorted.bam
```

## Apply to questions 15 - 19:
Compile candidate sites of variation using SAMtools mpileup for further evaluation with BCFtools. Provide the reference fasta genome and use the option “-uv” to generate the output in uncompressed VCF format for easy examination.

```console
$ samtools mpileup -uv -f wu_0.v7.fas align.full.sorted.bam -o align.full.mpileup.vcf
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
```

15. How many entries were reported for Chr3?
```console
$ cut -s -f1 align.full.mpileup.vcf | grep -c Chr3
360295
```

16. How many entries have ‘A’ as the corresponding genome letter?
```console
$ cut -s -f4 align.full.mpileup.vcf | grep -c "^A$"
1150985
```

17. How many entries have exactly 20 supporting reads (read depth)?
```console
$ cut -s -f8 align.full.mpileup.vcf | grep -c "DP=20;"
1816
```

18. How many entries represent indels?
```console
$ cut -s -f8 align.full.mpileup.vcf | grep -c "INDEL"
1972
```

19. How many entries are reported for position 175672 on Chr1?
```console
$ grep -c -P "^Chr1\t175672\t" align.full.mpileup.vcf
2
```

## Apply to questions 20 - 24:
Call variants using ‘BCFtools call’ with the multiallelic-caller model. For this, you will need to first re-run SAMtools mpileup with the BCF output format option (‘-g’) to generate the set of candidate sites to be evaluated by BCFtools. In the output to BCFtools, select to show only the variant sites, in uncompressed VCF format for easy examination.

```console
$ samtools mpileup -g -f wu_0.v7.fas align.full.sorted.bam -o align.full.bcf
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
$ bcftools call -v -m -O v -o align.full.vcf align.full.bcf
```

20. How many variants are called on Chr3?
```console
$ cut -s -f1 align.full.vcf | grep -c Chr3
398
```

21. How many variants represent an A->T SNP? If useful, you can use ‘grep –P’ to allow tabular spaces in the search term.
```console
$ cut -s -f4,5 align.full.vcf --output-delimiter="-" | grep -c "^A-T$"
392
```

22. How many entries are indels?
```console
$ cut -s -f8 align.full.vcf | grep -c "INDEL"
320
```

23. How many entries have precisely 20 supporting reads (read depth)?
```console
$ cut -s -f8 align.full.vcf | grep -c "DP=20"
2
```

24. What type of variant (i.e., SNP or INDEL) is called at position 11937923 on Chr3?
```console
$ grep -P "Chr3\t11937923" align.full.vcf
Chr3	11937923	.	G	A	13.73	.	DP=20;VDB=0.0587288;SGB=-0.556411;RPB=0.639909;MQB=0.931063;BQB=0.972484;MQ0F=0;ICB=1;HOB=0.5;AC=1;AN=2;DP4=16,0,4,0;MQ=39	GT:PL	0/1:48,0,137
# answer: SNP
```
